recipes/binutils/all/test_package/conanfile.py

```diff
@@ -1,11 +1,113 @@
+
+    @property
+    def _test_linker_args(self):
+        args = []
+        if self._target_os == "Windows":
+            args.extend(["--subsystem", "console", f"{self.build_folder}/lib/libkernel32.a"])
+        return args
+
+    @property
+    def _test_package_assembly_source(self):
+        part_arch = self._target_arch
+        if "armv7" in part_arch:
+            part_arch = "armv7"
+        elif part_arch in ("riscv32", "riscv64"):
+            part_arch = "riscv"
+        return os.path.join(self.source_folder, f"{self._target_os}-{part_arch}.s")
+
+    def _append_gnu_triplet(self, exe):
+        return f"{self.deps_user_info['binutils'].prefix}{exe}"
+
+    def build(self):
+        if not tools.cross_building(self):
+
+            if not os.path.isfile(self._test_package_assembly_source):
+                self.output.warn(f"Missing {self._test_package_assembly_source}.\ntest_package does not support this target os/arch. Please consider adding it. (It's a great learning experience)")
+            else:
+                tools.mkdir(os.path.join(self.build_folder, "bin"))
+                tools.mkdir(os.path.join(self.build_folder, "lib"))
+
+                gas = self._append_gnu_triplet("as")
+                ld = self._append_gnu_triplet("ld")
+                extension = ""
+                if self._target_os == "Windows":
+                    extension = ".exe"
+
+                    # Create minimum import library for kernel32.dll
+                    dlltool = f"{self.deps_user_info['binutils'].gnu_triplet}-dlltool"
+
+                    dlltool_args = [dlltool, "--input-def", f"{self.source_folder}/Windows-kernel32.def", "--output-lib", f"{self.build_folder}/lib/libkernel32.a"]
+                    self.run(" ".join(dlltool_args))
+
+
+                assembler_args = [gas, self._test_package_assembly_source, "-o", f"{self.build_folder}/object.o"]
+                linker_args = [ld, f"{self.build_folder}/object.o", "-o", f"{self.build_folder}/bin/test_package{extension}"] + self._test_linker_args
+
+                self.run(" ".join(assembler_args))
+                self.run(" ".join(linker_args))
+
+    def _can_run_target(self):
+        if self._settings_build.os != self._target_os:
+            return False
+        if self._settings_build.arch == "x86_64":
+            return self._target_arch in ("x86", "x86_64")
+        return self._settings_build.arch == self._target_arch
+
+    def _has_as(self):
+        if self._target_os in ("Macos"):
+            return False
+        return True
+
+    def _has_ld(self):
+        if self._target_os in ("Macos"):
+            return False
+        return True
 
     def test(self):
+        # Run selftest (conversion between conan os/arch <=> gnu triplet)
+        with tools.chdir(os.path.dirname(self.deps_user_info["binutils"].recipe_path)):
+            self.run(f"python -m unittest {os.path.basename(self.deps_user_info['binutils'].recipe_path)} --verbose")
```

Comment:
I agree testing these things is quite important. There is a lot of time spent there. I wonder if those test classes should be moved inside the `test_package/` instead of living in the root `conanfile.py`. Probably from those files we could import the root `conanfile` to reuse helper classes.

Of course, that could be a refactor for another PR.

Reply:
The self-test can be moved to the test package, if it becomes possible to export functions from a recipe to its dependencies.
In previous experiments, I could only export strings through `self.user_info`.

Reply:
I was not thinking about _exporting_ those functions, but about `from ..conanfile import XXXX` in `test_package/conanfile.py` (if it works, given we are using Py3).

Reply:
I did not choose that approach because I wanted to have the ability to test an exported recipe, not the recipe that happens to live in the cci repo.

So when you run the following command,
```
conan test recipes/binutils/all/test_package binutils/2.38@
```
you know the exported recipe is tested, not a combination of the exported recipe and the recipe in the cci file tree.

Reply:
for completeness sake, you can still run the self-tests of the file in the cci file tree without having to export/build anything, by running:
```bash
python -m unittest recipes/binutils/all/conanfile.py
```